docs: add zerofs comparison (#7366)
juicedata/juicefs: docs/en/introduction/comparison/juicefs_vs_zerofs.md

---
slug: /comparison/juicefs_vs_zerofs
description: This document compares ZeroFS and JuiceFS, examining their product positioning, architecture, and features.
---

# JuiceFS vs. ZeroFS

[ZeroFS](https://github.com/Barre/ZeroFS) is an open-source log-structured file system that turns S3-compatible buckets
into POSIX file systems, exposed over NFS and 9P, or as raw block devices over NBD. It keeps both file data and metadata
in the object store itself and relies on a local disk/memory cache for performance. ZeroFS is dual-licensed under AGPLv3
and a commercial license.

Both ZeroFS and [JuiceFS](https://github.com/juicedata/juicefs) turn object storage into a POSIX file system without
mapping one file to one object, but they differ substantially in metadata architecture, concurrency model, encryption
defaults, and licensing.

## Product positioning

**ZeroFS** is a self-hosted, encrypted-by-default file system for putting POSIX semantics (or a ZFS pool, or a block
device) on top of an S3-compatible bucket. A single ZeroFS core service handles metadata, data, and every protocol

> endpoint, so there's no separate metadata database to run alongside it. It targets teams that want to self-host

Reviewer: ```suggestion
endpoint, so there is no separate metadata database to run alongside it. It targets teams that want to self-host
```

which is similar to JuiceFS' approach in this regard.

Key architectural characteristics:

- No separate metadata database: metadata (inodes, directory entries, extent pointers, manifests) lives in an
  object-backed LSM tree, while file contents live separately in immutable segment objects, both managed by the ZeroFS
  server. When working with ZeroFS, this server (and its standby for HA and read replicas for read scaling) is the main
  service you deploy and operate.
- One server process (the leader) is the sole writer to the backing object store. It accepts connections from concurrent
  clients and arbitrates their writes and POSIX locks internally, but every mutation is funneled through that single
  process. See [Concurrency and write scalability](#concurrency-and-write-scalability) below.
- Encryption and compression are mandatory on every block, with no option to disable them.
- A dual-tier local cache (encrypted raw-parts cache for object bytes, plaintext decoded-block cache for metadata, plus
  an in-memory tail cache for each inode's most recent extent) reduces round trips to object storage.

> - For High availability, ZeroFS supports a leader/standby replication design with a single active writer and writer

Reviewer: ```suggestion
- For high availability, ZeroFS supports a leader/standby replication design with a single active writer and writer
```

**JuiceFS** adopts a decoupled architecture that separates metadata management from data storage. Files are split into
chunks (default up to 64 MiB, further divided into 4 MiB blocks) before being uploaded to object storage, with
corresponding metadata stored in a separate, pluggable database engine.

![JuiceFS-arch](../../images/juicefs-arch.svg)

Key architectural characteristics:

- JuiceFS supports pluggable metadata engines (Redis, TiKV, MySQL, PostgreSQL, SQLite, etcd, FoundationDB, etc.), each
  with its own operational and HA characteristics.
- JuiceFS stores files using data chunking, which enables efficient partial updates, appends, and high-throughput
  operations. For more details, see [Architecture](../../introduction/architecture.md).
- No dependency on any proprietary intermediate storage format. JuiceFS supports a flexible caching mechanism to
  decrease latency and improve performance.
- Multi-cloud and hybrid cloud support
  with [all mainstream object storage backends](../../reference/how_to_set_up_object_storage.md).

### Data path

**ZeroFS** relays all file data through its server node. Neither the NFS and 9P protocol clients nor the native Linux

> kernel client talk to object storage directly, only the ZeroFS server holds the storage credentials and issues the

Reviewer: ```suggestion
kernel client talk to object storage directly. Only the ZeroFS server holds the storage credentials and issues the
```

GET/PUT calls. So reads and writes are proxied through that one process on its way to or from the bucket.

**JuiceFS** clients, on the other hand, directly issue GET/PUT calls to the object store themselves. Metadata operations
go through the metadata engine, but file data always travels directly between the client mounting machines and the
object store, without an intermediate relay.

This distinction carries considerations in cloud deployments. Because ZeroFS clients never reach the object store
themselves, file data effectively crosses the network twice: once between the object store and the ZeroFS server, and

> again between the server and the client. When the server, its clients, and the object store don't all sit in the same

Reviewer: ```suggestion
again between the server and the client. When the server, its clients, and the object store do not all sit in the same
```

never a partial or torn write.

**JuiceFS** also buffers writes, in the client's read/write buffer (default 300 MiB), but ties the durability boundary
directly to `close()`: `close()`, `fsync()`, and `fdatasync()` (as well as filling a 4 MiB block) all trigger a flush,
and the client will not commit a write to the metadata engine until the corresponding data has actually been uploaded to
object storage. This is what its default "close-to-open" consistency guarantee is built on: a successful `close()`
already means the write is both durable in object storage and visible to any other client that opens the file afterward,
without the application ever having to call `fsync()` itself. For more details,
see [Cache](../../guide/cache.md#consistency) and
[POSIX compatibility](../../reference/posix_compatibility.md). JuiceFS does offer an opt-in `--writeback` mode with a
risk profile similar to ZeroFS's default: writes are committed to metadata immediately and uploaded from local disk
asynchronously. However, the docs flag it explicitly ("if write cache data suffers loss before upload is complete, file
data is lost forever") and it is off by default.

### Caching

> **ZeroFS** implements a dual-tier local cache split between disk and memory budgets (`disk_size_gb` / `memory_size_gb`

Reviewer: ```suggestion
**ZeroFS** implements a dual-tier local cache split between disk and memory budgets (`disk_size_gb`/`memory_size_gb`
```

## Licensing implications

**ZeroFS** is dual-licensed under the GNU AGPLv3 and a commercial license. The AGPLv3 can impose source-offering and
corresponding-source obligations when modified software is conveyed or made available for network interaction, subject
to the license's precise terms. Organizations should obtain legal advice for hosted, embedded, or distributed offerings.

**JuiceFS Community Edition** is released under the Apache License 2.0, a permissive license with no copyleft or
source-disclosure obligations for internal use, distribution, or hosted services. JuiceFS Enterprise Edition is
available under a separate commercial license for teams that need its distributed metadata engine, distributed cache,
and multi-cloud mirroring.

## Summary

**ZeroFS** is a security-first file system that folds metadata storage into its own server process instead of requiring
a separate database, encrypts and compresses every block by default, and can be mounted over NFS, 9P, or a native Linux

> kernel client, or consumed as an NBD block device. It's still a service you deploy and operate. You just don't operate a

Reviewer: ```suggestion
kernel client, or consumed as an NBD block device. It is still a service you deploy and operate. You just do not operate a
```
